src/wallet/wallet2_basic/wallet2_storage.cpp

```diff
@@ -0,0 +1,677 @@
+// Copyright (c) 2023, The Monero Project
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "wallet2_storage.h"
+
+#include <boost/archive/binary_iarchive.hpp>
+#include <boost/archive/portable_binary_iarchive.hpp>
+#include <rapidjson/document.h>
+
+#include "device/device_cold.hpp"
+#include "device_trezor/device_trezor.hpp"
+#include "file_io_utils.h"
+#include "serialization/binary_utils.h"
+#include "storages/portable_storage_template_helper.h"
+
+#define TRY_NOFAIL(stmt) try { stmt; } catch (...) {}
+
+#define READ_KEYS_JSON_FIELD_N(json, name, jtype, mandatory, var)     \
+    read_json_object_key_##jtype(var, json, #name, mandatory) \
+
+#define READ_KEYS_JSON_FIELD(json, name, jtype, mandatory)            \
+    READ_KEYS_JSON_FIELD_N(json, name, jtype, mandatory, kd.m_##name) \
+
+namespace
+{
+static hw::i_device_callback noop_device_cb;
+
+template <class Archive>
+wallet2_basic::cache boost_deserialize_cache(const std::string& cache_data)
+{
+    wallet2_basic::cache c;
+    std::istringstream iss(cache_data);
+    Archive ar(iss);
+    ar >> c;
+    return c;
+}
+
+template <typename T>
+void read_json_object_key_Int(T& out, const rapidjson::Value& json, const char* name, bool mand)
```

Comment:
Is all this new serialization stuff? Why rapidjson all of a sudden? I have a [msgpack implemenation](https://github.com/vtnerd/monero-lws/tree/develop/src/wire/msgpack) ready with [unit tests](https://github.com/vtnerd/monero-lws/blob/develop/tests/unit/wire/msgpack/read.write.test.cpp).

Reply:
I answered my own question eventually, this isn't just moving around files @UkoeHB this is changing the serialization format to a JSON format backed by rapidjson.

Reply:
My understanding is that this library is only intended to be used to read old wallet formats, so we can drop wallet2 while keeping support for old formats. None of this is new, .keys files use JSON today.

Reply:
@tobtoht is correct. This lib defines new types `cache` and `keys_data` and has analogous deserialization code to `wallet2::load`.  The generally point of the lib is to allow the core project to support the current wallet formats moving forwards, but ditch the vast majority of `wallet2` code.